### ISSUE TITLE
Future direction of BoxBilling

### ISSUE THREAD

**Issue by John-S4** [boxbilling/boxbilling#934]
### Discussed in https://github.com/boxbilling/boxbilling/discussions/933

<div type='discussions-op-text'>

<sup>Originally posted by **John-S4** August  7, 2021</sup>
I think that the direction of this software is currently a little bit unclear, and I am curious what other people actually expect or want from it. 

At the moment it seems to not really know what it wants to be. 

I guess essentially the question is whether people are looking for a big piece of software that does everything, or a light framework that does all of the core functions really well, and is easily extensible with additional 'modules',  'plugins', or whatever you choose to call them. 

Personally I would like to see it move towards the second of those options.

For example:

- I don't think that file manager belongs in there - I think that is something for a server control panel, not for a billing and client management portal. 

- I also don't think that the forum should be installed by default, if it is there as part of the core at all then it should definitely be an option on install. It's far from the best forum solution out there and I think it would be better to have nice integration with something better if you want a forum at all. 
 
It would be good to ship with a limited number of payment integrations, control panel integrations, etc. and have  a clear well-maintained repository of others that can be added. 

What are your views on what you want from BB and what direction the codebase should go in?</div>

**Comment by BenNottelling** [comment]
> * I don't think that file manager belongs in there - I think that is something for a server control panel, not for a billing and client management portal.
> *

Done as of last night ;)

Now, in my opinion BB needs to be a solid, stable core that provides the essentials for most people out of box, with the intent that extra features can be loaded in via extensions. 

Core:

- Billing, invoicing, and maybe ticketing
- Server management and control panel integration for a handful of popular control panels
  - Extensions available for others, these would probably be community maintained
- Well tested payment gateways that are maintained, VS the current option which is many, untested and unmaintained gateways
  - Thinking PayPal, Amazon Pay, and Stripe would be some good out of the box payment gateways
- Regular security and update patches to ensure the core is always a solid foundation people can grow upon

Optional:

- Forums
- Knowledge Base
- Possibly the ticketing system

Extensions: 
I believe that extensions should more-or-less be community maintained.  That doesn't mean we won't contribute, but in my opinion it's our job to provide a really solid core first and then anything else comes secondary. If an extension is part of someone 's business, I don't think it's unreasonable for them to pay a developer to maintain it, especially since most are probably integrations with pre-existing services or fairly simple, I don't think it would cost the person all that much money. 
Therefore, I feel that building and maintaining the core application while providing good documentation for anyone who wants to write their own extension should be enough in most cases. - This is probably the opinion that will get the most argument, but I'm happy to hear people's thoughts

**Comment by evrifaessa** [comment]
Hey guys, sorry that I don't have active time to dedicate to the development part. I have a working schedule that only keeps me some free time in the late evening, so I can't really help much for now.

Anyways, I also think we should make a bulk cleanup in the core. For example, as you also said, file manager shouldn't really be a part of the core of a "billing system". Keeping some unnecessary extensions a part of the core repository makes it hard for us -the core maintainers- to maintain, maybe we could just split them from the core, possibly to their own dedicated repositories for each extension, or remove some extensions at all.

Community should have a larger role at the extensions that core maintainers don't have time to maintain, and we should get the extension store up and running again. That way, it'll be easier to maintain the core.

To sum up, I agree with @BenNottelling and @John-S4. I'll try coming for help whenever I get to find some more free time. Good luck :)

**Comment by BenNottelling** [comment]
> we should get the extension store up and running again.

Agreed, I was thinking about this today and I think we could actually do this from the [extensions repository](https://github.com/boxbilling/extensions).
Thoughts:
- A master or development branch for testing and development.
- A stable branch for actual extension releases that are stable and bug free
  - This branch should include a "db" file file, perhaps in a json format
  - This file could include the name of every extension, the current version, a description, and URL for another json file that has all of the specific data for that extension.
    - Not completely sure what all info would be needed, but it should at least have a list of all files and appropriate location to update it from. Here's an example assuming the extension has 3 files. `core.php`, `request.php`, and `handle.hph`.
```
branches{
   stable{
     "version": "1.0.0",
     "files": "StableFiles"
   },
  development{
    "version": "dev",
    "files" "DevFiles"
  } 
}

StableFiles{
   "core.php": 'https://raw.githubusercontent.com/boxbilling/extensions/stable/example/core.php',
   "request.php": 'https://raw.githubusercontent.com/boxbilling/extensions/stable/example/request.php',
   "handle.php": 'https://raw.githubusercontent.com/boxbilling/extensions/stable/example/handle.php'
}

DevFiles{
   "core.php": 'https://raw.githubusercontent.com/boxbilling/extensions/master/example/core.php',
   "request.php": 'https://raw.githubusercontent.com/boxbilling/extensions/master/example/request.php',
   "handle.php": 'https://raw.githubusercontent.com/boxbilling/extensions/master/example/handle.php'
}
```

We could have BB pull all of these files straight from github which can help avoid issues like where we're at not with a broken boxbilling.com website and old installs that are still linked into it.
I'm sure I got the formatting wrong on that code example, but I feel like a solution similar to this could work.

Also, I'm glad to see you come by and be active, even if it's just for a bit :)

**Comment by Tenebras50** [comment]
I have tried few months back to put this project back on track poking people a bit to see if there was interest but havent received any positive confirmation as of yet, 

Think the problem as of now there aint a track of what need to be done. the priority and whos assigned to what if anyone

Without that organizing anything will be very difficult for anyone even if willing to contribute.

**Comment by Neustradamus** [comment]
@John-S4: Thanks for this issue, I have not seen before...

**Comment by John-S4** [comment]
> @John-S4: Thanks for this issue, I have not seen before...

I have not been around very much over the last months since I created this issue, I had planned to be a lot more, but have not been able to for personal reasons.

There is definite activity from other people though and the codebase is moving forward. I just think it all becomes a lot easier and clearer if there is an agreement about the direction that something is heading in, rather than everyone trying to do things in their own way and submitting pull requests that pull things in different directions and contradict each other.

**Comment by gOOvER** [comment]
> Absolutely agreed with topic starter. Project will be easier to develop if it becomes modular. But what's with technical part?
> 
> Some time I thinking how could be install Boxbilling extensions with `composer require organisation/repository`, but don't dig the code for that. Anybody have ready idea about this?
> 
> p.s. Project need the official roadmap!

Why using composer for plugins? I think this is not needed. In my eyes too big a security risk. Not 1 big project use composer for installing plugins & Addons

**Comment by jaapmarcus** [comment]
> What is really needed is first of all a community structure:
> 
> * Projectleader
> * Devs
> * Docu Team
> * Community Manager
>   
>   * Community Team

Please keep it small.. Unless you have have xx active contributors it makes no sense to set it up. Keep it small and flexible. Maybe in the future it is needed but currently not... 



> Why using composer for plugins? I think this is not needed. In my eyes too big a security risk. Not 1 big project use composer for installing plugins & Addons. An own Extesion System is better. I love the way, how Woltlab handle plugin/extension installs
> 
> I agree with you and the opinion of the creator. At the moment, everything is a bit confused because there are no clear structures.

Grav CMS, Magento all work with composer.. It has the advantages / disadvantages

**Comment by jaapmarcus** [comment]
> > > What is really needed is first of all a community structure:
> > > 
> > > * Projectleader
> > > * Devs
> > > * Docu Team
> > > * Community Manager
> > >   
> > >   * Community Team
> > 
> > 
> > Please keep it small.. Unless you have have xx active contributors it makes no sense to set it up. Keep it small and flexible. Maybe in the future it is needed but currently not...
> 
> Lesser people, means lesser support. Boxbilling have a, sry to say shit reputation, from the last owner. See how he act, when people talk about to remove that shit registration in an opensorce Project.
> 

I never have any experince with Boxbilling and personally don't use it my self.  So I don't really care what route the developers are going..

How ever I have experience with Hestia / Vesta and Vesta  happend the same thing until they decide to "drop" 1.0.0 and it is still broken.  Vesta was dead we had at least 100 / 200 switchers a month..  And we Hestia grew from 2k users 2 years ago to about 20k  currently. Even without an "Documentation" / Dev / Translations manager it still possible... 

There are issues:

1. There is a limited number of persons who are willing to invest a lot of time/money into a project 
2. Users that are available have limited time 
3. There is no backup for users if they fall out... for any reason there need to be users available who can take over the jobs.. 

I think the focus for Boxbilling need to be on a "stable" build first instead before rigging up a full "staff" 

And it makes sense to set a "point" so where Boxbilling need to "be" and it makes maybe a sense have a public roadmap for it..  Consider "removing" some current features as forum, FAQ and so on will not hurt... Boxbilling doesn't need to be a one trick pony that can be everything. It takes a lot of time to make it perfect for every thing... Keep in mind that maintainability in the future should have priority...

**Comment by Komodroid** [comment]
Are applications open? 👀

**Comment by BenNottelling** [comment]
> Are applications open? 👀

I mean, nobody is going to be paid anything but we welcome contributions

**Comment by m4rkstevenson** [comment]
I've used Boxbilling since the early days when Hosting24 were giving away free licences (I'm pretty sure, the same licence still in my unmodified bb-config file).. I'm glad to see some interest in developing this and I think the most important thing is providing a stable platform that supports the most common server panels and payment providers. I'm no dev, far from it but I'm committed to this project... mhosting.co.uk has been riding on it for years lol

**Comment by evrifaessa** [comment]
> I've used Boxbilling since the early days when Hosting24 were giving away free licences (I'm pretty sure, the same licence still in my unmodified bb-config file).. I'm glad to see some interest in developing this and I think the most important thing is providing a stable platform that supports the most common server panels and payment providers. I'm no dev, far from it but I'm committed to this project... mhosting.co.uk has been riding on it for years lol

Thanks for your input. Even if you're not interested in contributing to the source code, I'd love to see you in our Discord server.

We're trying to keep in touch with our old and new users and shaping the future of the project depending on their opinions. If you're also interested, click [here](https://boxbilling.org/discord). :)

**Comment by Mkoje** [comment]
So I would like to integrate some part of boxbilling on my web... I started a reseller hosting web and needed to just add the payment and  integrate it to work with my site... unfortunately boxbilling over rights every other content on my site changing my navbar, footer and showing all the boxbilling content... I have tried to find a way to edit all that out with no success. Can someone help?
